README.md
# Systemd Shim

fork of [systemd-shim](https://git.launchpad.net/ubuntu/+source/systemd-shim)


debian/changelog
systemd-shim (9-1bzr4ubuntu1) xenial; urgency=medium

  * debian/patches/dont-override-release-agent.patch: do not override an
    existing release agent.  If one is defined, something is going on that
    we don't know about, and we risk stopping user sessions immediately on
    login (preventing logins).

 -- Serge Hallyn <serge.hallyn@ubuntu.com>  Tue, 01 Mar 2016 20:57:52 -0800

systemd-shim (9-1bzr4) wily; urgency=medium

  * debian/patches/lp1390120.patch
    - Cherrypicked: Use the ntp support for 'systemd-timesyncd.service'
      too (LP: #1390120)

 -- Ken VanDine <ken.vandine@canonical.com>  Mon, 18 May 2015 10:43:32 -0400

systemd-shim (9-1bzr2) vivid; urgency=medium

  * autopkgtests: umount /sys/power/state instead of the source

 -- Iain Lane <iain@orangesquash.org.uk>  Thu, 19 Mar 2015 12:27:56 +0000

systemd-shim (9-1bzr1) vivid; urgency=medium

  * autopkgtests: The tests won't work when running under systemd; in that
    case, try to install upstart and reboot, and skip if that's not available.

 -- Martin Pitt <martin.pitt@ubuntu.com>  Wed, 11 Mar 2015 07:58:34 +0100

systemd-shim (9-1) unstable; urgency=medium

  * New upstream release.
    - Add org.freedesktop.systemd1.Manager Version property (Closes: #766237)
    - Clean up closing logind sessions (Closes: #756076)
  * Add myself to Uploaders: with Steve's consent.
  * Bump Standards-Version to 3.9.5 (no changes necessary).

 -- Martin Pitt <mpitt@debian.org>  Wed, 05 Nov 2014 21:55:33 +0100

systemd-shim (8-4) unstable; urgency=medium

  * Drop the dbus policy entirely from this package, as discussed in bug
    #765101; since the security policy should always be in sync with
    systemd's, and since the systemd package ships both logind (the consumer
    of systemd-shim) and this dbus policy, there's no reason to ship this
    separately rather than relying on the systemd copy.

 -- Steve Langasek <vorlon@debian.org>  Wed, 22 Oct 2014 04:29:44 +0000

systemd-shim (8-3) unstable; urgency=medium

  * debian/patches/dbus-policy-v215.patch: update our dbus policy to match
    that from systemd 215.  Closes: #765101.

 -- Steve Langasek <vorlon@debian.org>  Tue, 14 Oct 2014 04:54:09 +0000

systemd-shim (8-2) unstable; urgency=medium

  * No-change rebuild against systemd 215.

 -- Martin Pitt <mpitt@debian.org>  Tue, 16 Sep 2014 07:02:23 +0200

systemd-shim (8-1) unstable; urgency=medium

  * Update watch file.
  * New upstream release.
    - Drop patches cherry-picked from upstream.
  * Add versioned dependency on cgmanager (>= 0.32) for new APIs.

 -- Steve Langasek <vorlon@debian.org>  Wed, 10 Sep 2014 15:51:34 +0000

systemd-shim (7-3) unstable; urgency=medium

  * debian/patches/0001-cgmanager-stop-doing-async-calls.patch:
    cgmanager: stop doing async calls; fixes a race condition on login.
    Thanks to Ryan Lortie <desrt@desrt.ca>.  Closes LP: #1365095.

 -- Steve Langasek <vorlon@debian.org>  Fri, 05 Sep 2014 15:49:39 -0700

systemd-shim (7-2) unstable; urgency=medium

  * Fix Start TransientUnit signature for systemd >= 209. Check systemd
    version during package build. Patch from upstream git. (LP: #1351815)
  * Add systemd build dep for above configure check to work.
  * Generate Breaks: to systemd >= or << 209, depending on which version we
    build against.

 -- Martin Pitt <mpitt@debian.org>  Wed, 03 Sep 2014 08:07:00 +0200

systemd-shim (7-1) unstable; urgency=medium

  * New upstream release:
    - Provide cgmanager integration to work with logind >= 205.
    - Use "shutdown -h" instead of "poweroff" to respect /etc/default/halt
      (LP: #1193944)
    - Add stubs for Subscribe and Unsubscribe (LP: #1348039)
  * Drop all patches, included upstream.
  * debian/tests/lifecycle: Stop assuming that the shim isn't running
    initially, as it's now often activated during login, upgrade, etc.
    Instead, kill it manually to re-test D-BUS activation. Fixes autopkgtest
    regression.
  * debian/tests/power: Adjust to poweroff → shutdown upstream change.

 -- Martin Pitt <mpitt@debian.org>  Wed, 20 Aug 2014 08:22:08 +0200

systemd-shim (6-5) unstable; urgency=medium

  * Mark systemd-shim as Linux-only, since it now requires cgmanager which
    is also Linux-only.

 -- Steve Langasek <vorlon@debian.org>  Sun, 10 Aug 2014 22:30:43 -0700

systemd-shim (6-4) unstable; urgency=medium

  * Merge cgmanager support from Ubuntu; thanks to Serge Hallyn and
    Martin Pitt.  Closes: #752939.
  * Fix wrong path used in the .maintscript file.  Closes: #748791.

 -- Steve Langasek <vorlon@debian.org>  Fri, 25 Jul 2014 16:48:09 +0000

systemd-shim (6-3) unstable; urgency=medium

  [ Martin Pitt ]
  * Mark autopkgtest with "isolation-machine" as this test can't run in
    schroot or LXC.

  [ Steve Langasek ]
  * Sync /etc/dbus-1/system.d/org.freedesktop.systemd-shim.conf with
    /etc/dbus-1/system.d/org.freedesktop.systemd1.conf; since this only
    governs the security policy there's no reason for it to be different
    between the two implementations even though some of the declared
    interfaces will not be available on systemd-shim.  Closes: #746242.

 -- Steve Langasek <vorlon@debian.org>  Tue, 29 Apr 2014 07:49:10 +0000

systemd-shim (6-2) unstable; urgency=medium

  * Add diversions for the systemd package, and move the systemd-shim dbus
    file to a different filename under /etc/dbus-1 so that we can drop the
    Conflicts with systemd.

 -- Steve Langasek <vorlon@debian.org>  Wed, 01 Jan 2014 10:11:19 +0000

systemd-shim (6-1) unstable; urgency=medium

  * Initial upload to Debian.

 -- Steve Langasek <vorlon@debian.org>  Sat, 28 Dec 2013 21:05:01 +0000

systemd-shim (6-0ubuntu1) trusty; urgency=low

  * New upstream release:
    - Flush the right bus (system, not session), to address another case of
      "PrepareForSleep signal is missing" which breaks networking after
      resuming. (LP: #1252121)

 -- Martin Pitt <martin.pitt@ubuntu.com>  Tue, 10 Dec 2013 10:54:18 +0100


src/unit.c
/*
 * Copyright © 2013 Canonical Limited
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the licence, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
 * USA.
 */

#include "unit.h"

G_DEFINE_TYPE (Unit, unit, G_TYPE_OBJECT)

static void
unit_init (Unit *unit)
{
}

static void
unit_class_init (UnitClass *class)
{
}

Unit *
lookup_unit (GVariant  *parameters,
             GError   **error)
{
  const gchar *unit_name;
  Unit *unit = NULL;

  g_variant_get_child (parameters, 0, "&s", &unit_name);

  if (g_str_equal (unit_name, "ntpd.service"))
    unit = ntp_unit_get ();

  if (g_str_equal (unit_name, "suspend.target"))
    unit = power_unit_new (POWER_SUSPEND);

  else if (g_str_equal (unit_name, "hibernate.target"))
    unit = power_unit_new (POWER_HIBERNATE);

  else if (g_str_equal (unit_name, "reboot.target"))
    unit = power_unit_new (POWER_REBOOT);

  else if (g_str_equal (unit_name, "shutdown.target"))
    unit = power_unit_new (POWER_OFF);

  if (unit == NULL)
    g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FILE_NOT_FOUND,
                 "Unknown unit: %s", unit_name);

  return unit;
}

const gchar *
unit_get_state (Unit *unit)
{
  g_return_val_if_fail (unit != NULL, NULL);

  return UNIT_GET_CLASS (unit)->get_state (unit);
}

void
unit_start (Unit *unit)
{
  g_return_if_fail (unit != NULL);

  return UNIT_GET_CLASS (unit)->start (unit);
}

void
unit_stop (Unit *unit)
{
  g_return_if_fail (unit != NULL);

  return UNIT_GET_CLASS (unit)->stop (unit);
}


src/power-unit.c
/*
 * Copyright © 2013 Canonical Limited
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the licence, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
 * USA.
 */

#include "unit.h"

#include <stdlib.h>

typedef UnitClass PowerUnitClass;
static GType power_unit_get_type (void);

const gchar *power_cmds[] = {
  [POWER_OFF] = "/sbin/poweroff",
  [POWER_REBOOT] = "/sbin/reboot",
  [POWER_SUSPEND] = "/usr/sbin/pm-suspend",
  [POWER_HIBERNATE] = "/usr/sbin/pm-hibernate"
};

typedef struct
{
  Unit parent_instance;
  PowerAction action;
} PowerUnit;

G_DEFINE_TYPE (PowerUnit, power_unit, UNIT_TYPE)

static void
power_unit_start (Unit *unit)
{
  PowerUnit *pu = (PowerUnit *) unit;
  static gint64 last_suspend_time;
  static gboolean in_shutdown;

  /* If we request power off or reboot actions then we should ignore any
   * suspend or hibernate actions that come after this.
   */
  if (pu->action == POWER_OFF || pu->action == POWER_REBOOT)
    {
      in_shutdown = TRUE;
      system (power_cmds[pu->action]);
    }
  else
    {
      if (in_shutdown)
        return;

      /* This is pretty ugly: if we are being asked to perform a suspend
       * or hibernate action within 1 second of the previous one, don't
       * do it.
       *
       * We will be able to do this properly once we forward the
       * timestamp of the event that caused the suspend all the way
       * down.
       */
      if (pu->action == POWER_SUSPEND && last_suspend_time + G_TIME_SPAN_SECOND > g_get_monotonic_time ())
        return;

      system (power_cmds[pu->action]);

      if (pu->action == POWER_SUSPEND)
        last_suspend_time = g_get_monotonic_time ();
    }
}

static void
power_unit_stop (Unit *unit)
{
}

static const gchar *
power_unit_get_state (Unit *unit)
{
  return "static";
}

Unit *
power_unit_new (PowerAction action)
{
  PowerUnit *unit;

  g_return_val_if_fail (action < N_POWER_ACTIONS, NULL);

  unit = g_object_new (power_unit_get_type (), NULL);
  unit->action = action;

  return (Unit *) unit;
}

static void
power_unit_init (PowerUnit *unit)
{
}

static void
power_unit_class_init (UnitClass *class)
{
  class->start = power_unit_start;
  class->stop = power_unit_stop;
  class->get_state = power_unit_get_state;
}
